File: packages/php/src/Resources/Messages.php
<?php

declare(strict_types=1);

namespace Opensms\Resources;

use Opensms\Models;
use Opensms\Page;
use Opensms\Transport;

/**
 * The `messages` resource: send SMS, list and inspect messages, cancel scheduled ones.
 * Accessed as `$client->messages`.
 */
final class Messages
{
    /** @internal */
    public function __construct(private readonly Transport $transport)
    {
    }

    /**
     * Send one SMS. Returns the Message (201).
     *
     * Params: `to` (E.164, required), `text` (required), `sender_id`,
     * `traffic_type` (otp|transactional|marketing), `scheduled_at`
     * (DateTimeInterface or RFC 3339 string), `callback_url`, `metadata`.
     * camelCase aliases (`senderId`, ...) are accepted.
     *
     * @param array<string, mixed> $params
     * @param array{idempotencyKey?: string} $options
     * @return array<string, mixed>
     */
    public function send(array $params, array $options = []): array
    {
        $body = Models::map($params, Models::MESSAGE_SEND, ['to', 'text']);

        return Models::object($this->transport->request('POST', '/v1/messages', [], $body, [
            'idempotent' => true,
            'idempotencyKey' => Models::idempotencyKey($options),
        ]));
    }

    /**
     * List messages, newest first. Query: `limit` (1..100), `cursor`,
     * `status`, `to`, `country`, `date_from`, `date_to`.
     *
     * @param array<string, mixed> $params
     */
    public function list(array $params = []): Page
    {
        $query = Models::map($params, Models::MESSAGE_LIST_QUERY, [], true);

        return Models::page($this->transport->request('GET', '/v1/messages', $query));
    }

    /**
     * Fetch one message.
     *
     * @return array<string, mixed>
     */
    public function get(string $id): array
    {
        return Models::object($this->transport->request('GET', '/v1/messages/' . Models::pathParam($id)));
    }

    /**
     * List provider submission attempts for a message.
     *
     * @return list<array<string, mixed>>
     */
    public function attempts(string $id): array
    {
        return Models::list($this->transport->request('GET', '/v1/messages/' . Models::pathParam($id) . '/attempts'));
    }

    /**
     * Cancel a queued or scheduled message. Not retried automatically.
     *
     * @return array<string, mixed>
     */
    public function cancel(string $id): array
    {
        return Models::object($this->transport->request('POST', '/v1/messages/' . Models::pathParam($id) . '/cancel', [], null, [
            'retry' => false,
        ]));
    }
}

File: packages/java/src/main/java/io/opensms/Messages.java
package io.opensms;

import io.opensms.models.Message;
import io.opensms.models.MessageAttempt;
import io.opensms.models.Page;

import java.util.List;

/** Send and inspect SMS messages. Accessed as {@code client.messages()}. */
public final class Messages {

    private final ApiTransport t;

    Messages(ApiTransport t) {
        this.t = t;
    }

    /**
     * Send one SMS ({@code POST /v1/messages}). An {@code Idempotency-Key} is
     * generated when you do not pass one, and reused on retries.
     *
     * @param params the message.
     * @return the accepted message.
     */
    public Message send(SendMessageParams params) {
        return send(params, null);
    }

    /**
     * Send one SMS with per-call options.
     *
     * @param params  the message.
     * @param options for example an explicit idempotency key.
     * @return the accepted message.
     */
    public Message send(SendMessageParams params, RequestOptions options) {
        return t.json("POST", "/v1/messages", params.toWire(), ApiTransport.idempotencyKey(options),
                ApiTransport.type(Message.class));
    }

    /** @return the first page of messages, newest first. */
    public Page<Message> list() {
        return list(new MessageListParams());
    }

    /**
     * List messages ({@code GET /v1/messages}).
     *
     * @param params filters and paging.
     * @return one page.
     */
    public Page<Message> list(MessageListParams params) {
        return t.get("/v1/messages", params.toQuery(), ApiTransport.pageOf(Message.class));
    }

    /**
     * @param id message id.
     * @return the message.
     */
    public Message get(String id) {
        return t.get("/v1/messages/" + Wire.id(id, "id"), null, ApiTransport.type(Message.class));
    }

    /**
     * @param id message id.
     * @return delivery attempts, oldest first.
     */
    public List<MessageAttempt> attempts(String id) {
        return t.get("/v1/messages/" + Wire.id(id, "id") + "/attempts", null,
                ApiTransport.listOf(MessageAttempt.class));
    }

    /**
     * Cancel a queued or scheduled message. Never retried automatically.
     *
     * @param id message id.
     * @return the cancelled message.
     */
    public Message cancel(String id) {
        return t.json("POST", "/v1/messages/" + Wire.id(id, "id") + "/cancel", null, null,
                ApiTransport.type(Message.class));
    }
}
